pkg/cloud/ratelimit.go

```diff
@@ -112,3 +112,108 @@ func (m *MinimumRateLimiter) Accept(ctx context.Context, key *RateLimitKey) erro
 func (m *MinimumRateLimiter) Observe(ctx context.Context, err error, key *RateLimitKey) {
 	m.RateLimiter.Observe(ctx, err, key)
 }
+
+type TickerRateLimiter struct {
+	ticker *time.Ticker
+}
+
+func NewTickerRateLimiter(limit int, interval time.Duration) *TickerRateLimiter {
+	return &TickerRateLimiter{
+		ticker: time.NewTicker(interval / time.Duration(limit)),
+	}
+}
+
+// Accept everything immediately.
+func (t *TickerRateLimiter) Accept(ctx context.Context, rlk *RateLimitKey) error {
+	select {
+	case <-t.ticker.C:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Observe does nothing.
+func (*TickerRateLimiter) Observe(context.Context, error, *RateLimitKey) {
+}
+
+// CompositeRateLimiter combines rate limiters based on RateLimitKey.
+type CompositeRateLimiter struct {
+	rateLimiters map[string]map[string]map[string]RateLimiter
+	defaultRL    RateLimiter
+}
+
+// NewCompositeRateLimiter creates a new CompositeRateLimiter that will use
+// provided default rate limiter if no better match is found.
+func NewCompositeRateLimiter(defaultRL RateLimiter) *CompositeRateLimiter {
+	m := map[string]map[string]map[string]RateLimiter{
+		"": {
+			"": {
+				"": defaultRL,
+			},
+		},
+	}
+	return &CompositeRateLimiter{
+		rateLimiters: m,
+		defaultRL:    defaultRL,
+	}
+}
+
+func (c *CompositeRateLimiter) ensureExists(project, service string) {
+	if _, ok := c.rateLimiters[project]; !ok {
+		c.rateLimiters[project] = map[string]map[string]RateLimiter{}
+	}
+	if _, ok := c.rateLimiters[project][service]; !ok {
+		c.rateLimiters[project][service] = map[string]RateLimiter{}
+	}
+}
+
+func (c *CompositeRateLimiter) fillMissing() {
+	for _, subProject := range c.rateLimiters {
+		if subProject[""] == nil {
+			subProject[""] = map[string]RateLimiter{}
+		}
+		for _, subService := range subProject {
+			if subService[""] == nil {
+				subService[""] = c.defaultRL
```

Comment:
what is expected bahaviour here? defaultURL is a pointer so you copy pointer to all subServices. I would expect that you copy/create a new instance of RL so you can keep track request per Service and I don't think that this is working as expected. 
If you don't want separate RL instance for Service then this logic is too complex and you don't need to fill missing fields.

Reply:
Please add complex test which will cover this case.

Reply:
"keep track request per Service" - I don't understand. The goal is specifically to share rate limiters. Otherwise there will be a lot of issues with leaking resources (and growing resource use) for each path added.

"this logic is too complex" - the main goal of this function is to avoid complexity. By prefilling entries for empty keys we make `Accept` function significantly simpler

Reply:
My point is that you are `ensuring exists` and then `filling missing` where you just put default rate limiter to the map and later in n`Accept` function you check for each key in map and set default as empty string to get the RateLimiter.
I would expect that if you ensure that all rate limiters are in the map you don't do the check in Accept or just don't do the `ensure` and `filling` and return default RateLimitter if key does not exist.

This is the Accept function I'm referring to (added some comments inline): 
```
func (c *CompositeRateLimiter) Accept(ctx context.Context, rlk *RateLimitKey) error {
	if rlk == nil {
		return c.defaultRL.Accept(ctx, rlk)
	}
	project := rlk.ProjectID
	if _, ok := c.rateLimiters[project]; !ok {
		project = ""  // You are calling fillMissing function but still there is a need to change project to default?
	}
	service := rlk.Service
	if _, ok := c.rateLimiters[project][service]; !ok {
		service = ""
	}
	operation := rlk.Operation
	if _, ok := c.rateLimiters[project][service][operation]; !ok {
		operation = "" // you don't ensure/fill missing operation as project + service
	}
	return c.rateLimiters[project][service][operation].Accept(ctx, rlk)
}
```

So either simplify Accept function because you rely on Register function and verify rl settings like project once at the beginning or return default whenb no project is defined.

Reply:
> You are calling fillMissing function but still there is a need to change project to default?

Yes, that's the point.

> you don't ensure/fill missing operation as project + service

I do. In exactly the line that this comment is pinned to.

> I would expect that if you ensure that all rate limiters are in the map you don't do the check in Accept

Please provide a more comprehensive code snippet that would satisfy the following requirements:
1. Setting default rate limiter for a project
2. Setting default rate limiter for a service
3. Setting default rate limiter for an operation
4. Setting default rate limiter for service+operation combination

Reply:
ok, that sounds reasonable. There are two options I see:
1. remove project from the path specification (rate limiters will be shared across all projects)
2. refactor this to use "rate limiter factory" instead of a rate limiter instances (this would require allocating a lot of stuff for the first call for each new project)

Reply:
Implemented 1 in this PR. I started working on 2 in https://github.com/AwesomePatrol/k8s-cloud-provider/compare/add-new-rate-limiter-types...AwesomePatrol:k8s-cloud-provider:add-per-project-rate-limiter?expand=1